python/private/config_settings.bzl

```diff
@@ -71,23 +76,26 @@ def is_python_config_setting(name, *, python_version, match_extra = None, **kwar
     Args:
         name: name for the target that will be created to be used in select statements.
         python_version: The python_version to be passed in the `flag_values` in the `config_setting`.
-        match_extra: The labels that should be used for matching the extra versions instead of creating
-            them on the fly. This will be passed to `config_setting_group.match_any`.
+        match_any: The labels that should be used for matching the extra versions instead of creating
+            them on the fly. This will be passed to `config_setting_group.match_any`. This can be
+            either None, which will create config settings necessary to match the `python_version` value,
+            a list of 'config_setting' labels passed to bazel-skylib's `config_setting_group` `match_any`
+            attribute.
         **kwargs: extra kwargs passed to the `config_setting`
     """
     visibility = kwargs.pop("visibility", [])
 
     flag_values = {
-        _PYTHON_VERSION: python_version,
+        _PYTHON_VERSION_FLAG: python_version,
     }
     if python_version not in name:
-        fail("The name must have the python version in it")
+        fail("The name '{}' must have the python version '{}' in it".format(name, python_version))
 
-    match_extra = match_extra or {
-        "_{}".format(name).replace(python_version, x): {_PYTHON_VERSION: x}
-        for x in VERSION_FLAG_VALUES[python_version]
-    }
-    if not match_extra:
+    if python_version not in VERSION_FLAG_VALUES:
+        fail("The 'python_version' must be known to 'rules_python', choose from the values: {}".format(VERSION_FLAG_VALUES.keys()))
+
+    python_versions = VERSION_FLAG_VALUES[python_version]
```

Comment:
In trying to understand the logic, I tried to rewrite it in a way that more directly represented what the doc said. The doc basically says, given python_version=X, and:
* Given match_any=None, then match any of the `3.X.*` states
* Given match_any=<some list>, then match of the `[3.X] + match_any` states (note no `.*` in this case). (Ah, actually my understanding on this part might be wrong...)

So, the two things we have to keep track of are (1) what do we have to create (create_config_settings)? and (2) what do we pass to config_setting_group.match_any (match_any)?

So, what about this:

```
if match_any == None:
  create_config_settings = {<build from VERSION_FLAG_VALUES[pythonn_version]}
  match_any = list(create_config_settings.keys())
else:
  create_config_settings = {_PYTHON_VERSION_FLAG: python_version}
  match_any = ...er, wait...

if len(match_any) == 1:
  assert create_config_settings is empty?? # ...er, wait ...
  native.config_setting(flag_values = match_any?)
else:
  for ... in create_config_settings: ...

  config_setting_group(match_any = match_any)
```

Er -- wait, what happened in the else-branch that assigned match_any? It looks like if match_any is used, then the python_version setting isn't part of the condition? I think that's intentional from what I vaguely recall of the use case motivating this PR. That wasn't clear from the doc and implementation, though. But also a bit weird? It means you end up passing python_version="3.8", but that variable isn't actually used besides (1) checking the name arg and (2) that the key exists in the VERSION_FLAG_VALUES dict, but no subsequent logic appears to rely on that?

And so, if we strip all that logic away, doesn't passing match_any effectively just make this a wrapper around calling `config_setting_group(match_any=match_any)` ?

Am I missing something?

Reply:
Yeah, I think there was an issue with the concepts being used here. I reworked the code and I think it makes it sligtly clearer.

Thanks for raising this concern.